### ISSUE TITLE
Computer knows the mountains without geologists

### ISSUE THREAD

**Issue by Spikeone** [Return-To-The-Roots/s25client#139]
Computer builds the mines, and sends out the geologists ONLY after it. Logically this should be reversed.
He can get big advantages on a new map.
Or this feature could be connected to different AI-types in the future - computer would know the mountains only in Harder difficulties.

Spike (fvv1) wrote on 2010-04-20:	#1
I think this was discussed already and it's not really bad for an AI that this works like it does. Do you really lose vs AI? so I think it's not a big advantage. and very hard to realize as jh said one time i think

Tamas Fonyodi (almoskapitany) wrote on 2010-04-20:	#2
Well, I'm new to this forum/board, please excuse me. And I'm not fanatic, probably close to a lame according to you. And yes, the current AI beats me on new maps, or it leads to crash.
I think in the future it would be definitely good to have more AI types for "lames" like me, too.

https://bugs.launchpad.net/s25rttr/+bug/567378

**Comment by MarcusSt** [comment]
I would not exactly call this a bug. It's rather a measure for the AI to keep up with really good players. Support for different AI strengths has been implemented.

**Comment by Spikeone** [comment]
More like a visual bug - maybe AI could build after the first message with "geologist found xy" and then the job would be executed?

**Comment by Flamefire** [comment]
Currently logic for sending geologists is not implemented. The ones seen are actually around the built mine. I added some logic so the AI gets notified about resources found so it **could** react. But that would mean that it needs some logic to place flags and call geologists first.
--> Postpone till e.g. @jhkl got some time

**Comment by jhkl** [comment]
Suggestion: Couple behaviour with the difficulty setting:
A "hard" AI knows the map, so it can instantly build mines on a mountain. This would be like a human player that knows the map from a previous game.
Medium/Easy have to send geologists and only build mines on the placed signs. For that we need a good heuristic to decide
* where to place a flag/multple flags on a mountain? Maybe just a single flag in the centroid of the intersection of mountain and own territory would suffice if this is done incrementally with each territory extension
* how many geologists to send? Fixed number? Based on mountain size? (Might be bad as geologists also place signs on non mountain terrain)
* when to actually place the mines? Somehow identify 'good' clusters of signs? Would be again more complicated. Wait until all geologist are done and return to the warehouse? Might take a lot of time. Nevertheless I would suggest waiting for all geologists to finish, remove the flag/road and then place the mines. If combined with the difficulty setting it seems to be okay to make the AI slower in building mines.

**Comment by Flamefire** [comment]
I agree with the hard setting.

Medium could be: Set a flag where the mine flag would be (as determined by hard), call geologists and build the mine once a sign has been placed at the (pre-determined) mine position.
Easy: Place a flag somewhere (however this is done) and call geologists. For every sign place search for optimal mine position and place mine on flag once the optimal position is close enough or equal

I refactored the resource maps of the AI a bit. However there still 2 usages (Check the FindBestPosition implementations in the AI and the resource map, the one form the res map is never used, so this might need a redesign first so it is at least consistent)

**Comment by Spikeone** [comment]
Sounds Good.
But medium AI should not wait till a geologist examines the exact spot where the AI wants to build a mine, 1 node away is also enough for an order.
If I understood correctly how AI handles mines, the easy AI could start with an empty resource map and add a job to examine unknown (not empty) mountain spots.

**Comment by Flamefire** [comment]
right, this is what I meant with "if the sign is close enough"

Resource maps are not really used like they should ATM. Their usage should be refactored... Currently the AI searches for the "best position" based on current resources as on the map. The easy AI could use the same search but only around an already placed sign

**Comment by Spikeone** [comment]
But the AI should remember placed signs for later like players do.
Sorry if I missunderstood you.

**Comment by Flamefire** [comment]
Sure about that? When a sign vanished, the resource can be already mined and is no longer available. And "remembering" does also not really work for human players in "normal" maps, where you have different resources on the same mountain. At least I can't except for easy places (e.g. a place between my coal mines is probably coal too)
Remember that this are the easy-medium AIs. If they want to place a mine, they (can) use a geologist as a "delay"

**Comment by Spikeone** [comment]
Well mediums player also add mines as ace holders - which is the same as an AI remembering spots.
Also the AI knows where nothing is because the mine is depleted.
Maybe the AI missplaces mines from time to time, just like a player does.
